clang/lib/Frontend/InitPreprocessor.cpp

```diff
@@ -1165,6 +1165,8 @@ static void InitializePredefinedMacros(const TargetInfo &TI,
   DefineType("__WCHAR_TYPE__", TI.getWCharType(), Builder);
   DefineType("__WINT_TYPE__", TI.getWIntType(), Builder);
   DefineTypeSizeAndWidth("__SIG_ATOMIC", TI.getSigAtomicType(), TI, Builder);
+  if (LangOpts.Char8 || LangOpts.C23)
+    DefineType("__CHAR8_TYPE__", TI.UnsignedChar, Builder);
```

Comment:
This matches GCC behaviour: https://godbolt.org/z/6Eax3eqrd

But I'm not sure why it's defined in C++ at all

Reply:
Hmmm that is surprising, because it expands to `unsigned char` and not `char8_t` there: https://godbolt.org/z/Y96bPbj8e and this seems to have changed between GCC 8 and GCC 9: https://godbolt.org/z/EaEPKr59E

It may be worth opening a GCC bug to see if this was intentional or not.

Reply:
According to https://gcc.gnu.org/legacy-ml/gcc-patches/2018-11/msg00295.html `__CHAR8_TYPE__` is "the underlying type of `char8_t`" in C++. `char8_t` in C++ was added in GCC9 which explains the difference between GCC8 and 9

And this made me realise `-std=c++11 -fchar8_t` sets `LangOpts.Char8` in C++11, which I think this patch handles fine (See also: #97601).
Also `-std=c23 -fchar8_t` is a mode that exists. It makes `typeof(u8'x')` `char8_t` and `typeof(u8"")` `char8_t[1]`, even with this patch, which I guess is fine. Pending #55373 if `-std=c23 -fchar8_t` even makes sense.

Reply:
I'm certainly concerned with the impact of that in C++. It should not be defined in C++ (ie, char9_t is not unsigned char in C++)

Reply:
FWIW, I think https://github.com/llvm/llvm-project/issues/55373 should be fixed so that the option is rejected in C modes, but that's somewhat orthogonal to this patch.